### PR TITLE
feat(kousa): implemented ip address hashing, fix tests, etc.

### PR DESCRIPTION
**Pull request by saihnii4**
this pr implements the hashing of ip addresses that are stored in the database and fixes kousa tests.

implements #2579

### CODE REVIEW

**Comment by benawad** [comment]
I think I want to use a faster hash than argon

**Comment by saihnii4** [comment]
how about bcrypt?

**Review by benawad — CHANGES_REQUESTED** [review]
switch to HMAC

**Comment by amitojsingh366** [comment]
please don't use npm on kebab, we use yarn for almost all of our workspaces. also delete the `package-lock.json`

File: kebab/package-lock.json [hunk 1]
```diff
@@ -0,0 +1,5716 @@
+{
```
Comment (amitojsingh366) [inline kebab/package-lock.json]: please don't use npm on kebab, we use yarn for almost all of our workspaces. also delete the `package-lock.json`

**Review by amitojsingh366 — CHANGES_REQUESTED** [review]

**Review by overlisted — COMMENTED** [review]
i only review kebab

File: kousa/Session.vim [hunk 2]
```diff
@@ -0,0 +1,118 @@
+let SessionLoad = 1
```
Comment (samarmohan) [inline kousa/Session.vim]: remove this file

**Review by samarmohan — APPROVED** [review]

File: kibbeh/package.json [hunk 3]
```diff
@@ -30,6 +30,7 @@
     "@material-ui/pickers": "^3.3.10",
     "@svgr/webpack": "^5.5.0",
     "@tailwindcss/line-clamp": "^0.2.0",
+    "bufferutil": "^4.0.3",
```
Comment (benawad) [inline kibbeh/package.json]: remove unrelated code

File: kousa/config/dev.exs [hunk 4]
```diff
@@ -31,6 +31,8 @@ config :kousa,
   api_url: System.get_env("API_URL") || "http://localhost:4001",
   secret_key_base:
     "213lo12j3kl21j3kl21alaksjdklasjdklajsldjsaldjlasdlaksjdklasjdklajsldjsaldjlasdlaksjdklasjdklajsldjsaldjlasdadjlasjddlkijoqwijdoqwjd12loki3jhl12jelk12jekl1221099dj120",
+  ip_hashing_key:
```
Comment (benawad) [inline kousa/config/dev.exs]: remove this

File: kousa/lib/kousa/utils/hash.ex [hunk 5]
```diff
@@ -1,21 +1,29 @@
...
 
   @spec hash_ip(String.t()) :: hash_result
   def hash_ip(ip) do
-    Argon2.hash_pwd_salt(ip)
+    if is_nil(ip) do
+      {:error, "ip is nil"}
+    else
+      # Should we refactor this so we don't pipe a function call?
+      # Encodes Hash into Base16
+      :crypto.mac(:hmac, :sha256, Application.fetch_env!(:kousa, :ip_hashing_key_base), ip)
```
Comment (ChrisDesigns) [inline kousa/lib/kousa/utils/hash.ex]: This isn't a good implementation.  It would be very easy to get the base key and create a rainbow table for the whole IP space. This methodology uses ip_hashing_key_base as "pepper".  To do this method properly, it should be done in conjunction with salts.  Looking at [RFC7518](https://datatracker.ietf.org/doc/html/rfc7518#section-3.2), the uniquely created salt should two time the length of the input (this would  mean 2*45 = 90 characters). The hash can be stored in the same field as the IP using the [crypt](https://en.wikipedia.org/wiki/Crypt_(C)) format. A new function could be created to handle the splitting of these two fields and use secure_compare to return if the IP + salt = hash

**Review by willdoescode — CHANGES_REQUESTED** [review]
Remove Session.vim file

**Comment by samarmohan** [comment]
> Remove Session.vim file

Lol I already told him to do that

File: kousa/lib/kousa/utils/hash.ex (hunk 5 above)
Comment (benawad) [reply kousa/lib/kousa/utils/hash.ex]: I thought the whole point is someone wouldn't have the hash key because its a secret

File: kousa/Session.vim (hunk 2 above)
Comment (saihnii4) [reply kousa/Session.vim]: done

File: kousa/lib/kousa/utils/hash.ex (hunk 5 above)
Comment (ChrisDesigns) [reply kousa/lib/kousa/utils/hash.ex]: The key should not be a substitute for a good salt. I think that this approach shares the same issues as a shared [pepper](https://en.wikipedia.org/wiki/Pepper_(cryptography)#Shared_Secret_Pepper).
Looking at it from the prospective of a hacker or an evil company, this value is store via environment variables `System.get_env("IP_HASHING_KEY_BASE")`. Being that value is constant (shared for all hashing) and can be fetched by both a company or hacker if they gain access to the system. It won't take much more work for either entity to create a rainbow table using this know secret (presuming they managed to get database access). Every IP is bijective with the entries of a rainbow table for a know key. The number of IP that exists in the database, denoted as **_n_**, is only a subset of the know rainbow table entity. So let's take all the Class C IPs, denoted as **_c_**, to create a valid rainbow table, the IP space would need to be computed for each salt. This will take the complexity from O(**_c_**) to O(**_c_*****_n_**). Don't get me wrong, I think having anything is better than no hashing. I just think the approach could be broken based on  the methods displayed.

Comment (saihnii4) [reply kousa/lib/kousa/utils/hash.ex]: how about generating and encrypting a secret private key using a encryption key and storing that private key in the database alongside the user? this method forces an attacker to get access to the database **and** the environment variables.

**Comment by saihnii4** [comment]
closing because stale ig
